### PR TITLE
clock: explain how time works

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by matklad** [comment]
cc @owickstrom

File: src/vsr/clock.zig [hunk 1]
```diff
@@ -1,3 +1,84 @@
...
+//! problem, the primary just takes the max between the current time and the previous timestamp
+//! used. Notably, this ends up preserving monotonicity across restarts --- it is when replaying
+//! past prepares from the WAL that a replica learns about the latest timestamp before restart.
+//!
+//! Third, replica's system time lacks high availability: if a primary is isolated from NTP servers
+//! its local clock can drift significantly. Another problematic scenario is an operator error
+//! which incorrectly adjusts primary's local clock to be far in the future, which, due to
+//! monotonicity requirement, could render the cluster completely unusable.
+//!
+//! To solve the last problem, primary aggregates clock information from the entire cluster and
```
Comment (sentientwaffle) [inline src/vsr/clock.zig]:
```suggestion
//! To solve the last problem, the primary aggregates clock information from the entire cluster and
```

**Review by sentientwaffle — DISMISSED** [review]
Excellent! The clear explanation of the algorithm in particular really helped my understanding.

File: src/vsr/clock.zig [hunk 2]
```diff
@@ -1,3 +1,84 @@
...
+//! Time plays a central role in TigerBeetle data model. Because it is so important, TigerBeetle
+//! defines its own time. In other words, we don't use time to drive consensus, we use consensus to
+//! drive time!
+//!
+//! Time is important for the domain of accounting (e.g., pending transfers can expire with time),
+//! but it can't be supplied by the client, as its clock can be unreliable. For this reason,
+//! TigerBeetle needs to expose a "time service" to the state machine logic.
+//!
+//! Additionally, TigerBeetle needs to assign some kind of a sequence number to every event in the
+//! system, to make it easy to say whether A happened before B or vice versa.
```
Comment (owickstrom) [inline src/vsr/clock.zig:12]: Worth pointing out that it's a total order and not partial?

File: src/vsr/clock.zig [hunk 3]
```diff
@@ -1,3 +1,84 @@
...
+//!
+//! The ultimate source of timestamps is each replica's operating system. This time is backed by a
+//! replica-local drifty hardware clock which is periodically synchronized through NTP with high
+//! quality clocks elsewhere. Using system time directly as a source of TigerBeetle timestamps
+//! doesn't work:
+//!
+//! First, system time differs across replicas. To solve this problem, only the primary assigns
+//! timestamps. Specifically, when the primary converts a request to a prepare, it assigns its
+//! current time to the prepare. The state machine then assigns `prepare_timestamp + object_index`
+//! as the creation timestamp for each object in a batch.
```
Comment (owickstrom) [inline src/vsr/clock.zig:33]: Would be good to note resolution here, i.e. that `prepare_timestamp` is in nanoseconds.

File: src/vsr/clock.zig [hunk 4]
```diff
@@ -1,3 +1,84 @@
...
+//!
+//! Second, system time is not monotonic: due to NTP it can easily go backwards. To solve this
+//! problem, the primary just takes the max between the current time and the previous timestamp
+//! used. Notably, this ends up preserving monotonicity across restarts --- it is when replaying
+//! past prepares from the WAL that a replica learns about the latest timestamp before restart.
```
Comment (owickstrom) [inline src/vsr/clock.zig:38]: This is mostly a question for my learning, not trying to poke holes: is there a possibility of an empty WAL at restart, and that there's no _latest timestamp_ to be found? Fallback to latest snapshot timestamp? (I'm obviously out of my depth here, but curious nonetheless.)

**Review by owickstrom — COMMENTED** [review]
Really good intro!

File: src/vsr/clock.zig (hunk 2 above)
Comment (matklad) [reply src/vsr/clock.zig:12]: I think it's a part of the general context that everything is strictly serializable, so it's not worth it here

File: src/vsr/clock.zig (hunk 3 above)
Comment (matklad) [reply src/vsr/clock.zig:33]: This is mentioned above:

```
each object in TigerBeetle gets tagged with a u64
nanosecond-precision creation timestamp
```

**Review by sentientwaffle — APPROVED** [review]

File: src/vsr/clock.zig (hunk 3 above)
Comment (owickstrom) [reply src/vsr/clock.zig:33]: Ah, missed it.

File: src/vsr/clock.zig (hunk 4 above)
Comment (matklad) [reply src/vsr/clock.zig:38]: I was actually curious about this myself, that's a bit tricky!

First, as a bit of terminology, the thing we have here is called checkpoint, not a snapshot (we also have snapshots, but they are a different thing).

And, surprisingly, checkpoint _doesn't_ directly include a timestamp.

But the tricky thing about the tigerbeetle is that checkpoints are staggered --- roughly, checkpoint that includes the state up to `op` gets created and written to disk only when the WAL proceeds up to `op + 32`.

So, it is indeed the case that the replica always have a checkpoint and at least one prepare on top of it in the WAL, and it is the prepare that carries over the timestamp.

Comment (owickstrom) [reply src/vsr/clock.zig:38]: Gotcha. In other words, unless the database is empty, the WAL is never empty?

**Comment by matklad** [comment]
```
test
+- test:integration
   +- run test
      +- zig test Debug x[8](https://github.com/tigerbeetle/tigerbeetle/actions/runs/10856143117/job/30130219645?pr=2310#step:4:9)6_64-linux
         +- options
            +- run build_multiversion (tigerbeetle) stderr
$ /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/08adb0e11a4562818615e22a0fc026dc/llvm-objcopy-x86_64-linux --enable-deterministic-archives /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/565e[9](https://github.com/tigerbeetle/tigerbeetle/actions/runs/10856143117/job/30130219645?pr=2310#step:4:10)c608ae6227e5c9ac89d69416389/tigerbeetle /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/tigerbeetle-working
$ /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/08adb0e11a4562818615e22a0fc026dc/llvm-objcopy-x86_64-linux --enable-deterministic-archives --keep-undefined --remove-section .tb_mvb --remove-section .tb_mvh /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/dd46a7f2fab285ae856c1925530fac4f/stdout /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/tigerbeetle-past-0.16.1-x86_64
$ /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/08adb0e11a4562818615e22a0fc026dc/llvm-objcopy-x86_64-linux --enable-deterministic-archives --keep-undefined --add-section .tb_mvb=/home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/multiversion.body --set-section-flags .tb_mvb=contents,noload,readonly --add-section .tb_mvh=/home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/multiversion-zero.header --set-section-flags .tb_mvh=contents,noload,readonly /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/tigerbeetle-working
$ /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/08adb0e11a4562818615e22a0fc026dc/llvm-objcopy-x86_64-linux --enable-deterministic-archives --keep-undefined --remove-section .tb_mvh --add-section .tb_mvh=/home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/multiversion.header --set-section-flags .tb_mvh=contents,noload,readonly /home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/tmp/6192417250555817603/tigerbeetle-working
test
+- test:integration
   +- run test 15/16 passed, 1 failed
error: 'integration_tests.test.benchmark/inspect smoke' failed: /home/runner/work/tigerbeetle/tigerbeetle/zig/lib/std/testing.zig:540:14: 0x[10](https://github.com/tigerbeetle/tigerbeetle/actions/runs/10856143117/job/30130219645?pr=2310#step:4:11)8b36f in expect (test)
    if (!ok) return error.TestUnexpectedResult;
             ^
/home/runner/work/tigerbeetle/tigerbeetle/src/integration_tests.zig:263:9: 0x1091d30 in test.benchmark/inspect smoke (test)
        try std.testing.expect(status_ok_inspect);
        ^
error: while executing test 'testing.snaptest.decltest.equal_excluding_ignored', the following test command failed:
/home/runner/work/tigerbeetle/tigerbeetle/.zig-cache/o/4016dd6f390de43cd0969df41c3a9e8c/test --listen=- 
Build Summary: 18/21 steps succeeded; 1 failed; 237/238 tests passed; 1 failed (disable with --summary none)
test transitive failure
+- test:integration transitive failure
   +- run test 15/16 passed, 1 failed
error: the following build command failed with exit code 1:
```

This sounds like at least two funs:

* Fun A) this clearly shouldn't be failing randomly
* Fun B) perhaps it's time to take yet another look at shell.zig's API --- the whole purpose of that module is to make debugging process failures easier because you see what fails where, and we are clearly violating this here

File: src/vsr/clock.zig (hunk 4 above)
Comment (matklad) [reply src/vsr/clock.zig:38]: Even for an empty database we have a root prepare in the WAL!

Comment (owickstrom) [reply src/vsr/clock.zig:38]: I need to read up on that. But yeah, that's an interesting invariant. Perhaps already enforced with assert? (On my phone and can't easily check right now)

**Comment by matklad** [comment]
Fun B is being had at #2316

**Pull request merged**